crates/matrix-sdk/src/event_cache/linked_chunk.rs

```diff
@@ -43,48 +38,156 @@ pub enum LinkedChunkError {
     InvalidItemIndex { index: usize },
 }
 
+/// Represent the updates that have happened inside a [`LinkedChunk`].
+///
+/// To retrieve the updates, use [`LinkedChunk::updates`].
+///
+/// These updates are useful to store a `LinkedChunk` in another form of
+/// storage, like a database or something similar.
+#[derive(Debug, PartialEq)]
+pub enum LinkedChunkUpdate<Item, Gap> {
+    /// A new chunk of kind Items has been created.
+    NewChunkItems {
+        /// The identifier of the previous chunk of this new chunk.
+        previous: Option<ChunkIdentifier>,
+        /// The identifier of the new chunk.
+        new: ChunkIdentifier,
+        /// The identifier of the next chunk of this new chunk.
+        next: Option<ChunkIdentifier>,
+    },
+    /// A new chunk of kind Gap has been created.
+    NewChunkGap {
+        /// The identifier of the previous chunk of this new chunk.
+        previous: Option<ChunkIdentifier>,
+        /// The identifier of the new chunk.
+        new: ChunkIdentifier,
+        /// The identifier of the next chunk of this new chunk.
+        next: Option<ChunkIdentifier>,
+        /// The content of the chunk.
+        gap: Gap,
+    },
+    /// A chunk has been removed.
+    RemoveChunk(ChunkIdentifier),
+    /// Items are inserted inside a chunk of kind Items.
+    InsertItems {
+        /// [`Position`] of the items.
+        at: Position,
+        /// The items.
+        items: Vec<Item>,
+    },
+    /// A chunk of kind Items has been truncated.
+    TruncateItems {
+        /// The identifier of the chunk.
+        chunk: ChunkIdentifier,
+        /// The new length of the chunk.
+        length: usize,
+    },
+}
+
+/// Links of a `LinkedChunk`, i.e. the first and last [`Chunk`].
+///
+/// That's an internal type to make the borrow checker happy. Indeed, to add
+/// safety around these `NonNull`s, we want to use the borrow checker as much as
+/// possible. For example, `Self::latest_chunk_mut` takes `&mut self` and
+/// returns `&mut Chunk`, which is great and safe and all. Except that, if the
+/// `first` and `last` fields are used directly inside [`LinkedChunk`], the
+/// borrow checker considers that `self` is entirely used mutably. Thus, when we
+/// use another field of `LinkedChunk` mutably or immutably, it violates the
+/// mutability rules. The borrow checker is smart though: If we extract some
+/// values in a particular type, which are placed in a single field, the borrow
+/// checker knows that the other fields are safe to use mutably or immutably.
+/// The checks are per-field, not per-struct. That's why `first` and `last` have
+/// been extracted in a separate type, this type.
+struct LinkedChunkLinks<const CHUNK_CAPACITY: usize, Item, Gap> {
```

Comment:
Sorry, I know I'm the one who suggested this name, but I misunderstood the content: since these are the ends of the linked list, maybe `LinkedChunkEnds` instead?

Reply:
I'm fine with both!